### PR TITLE
zhangqiu/polar

### PR DESCRIPTION
**Pull request by NeosZhang**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation and Context

add polar

## Description

add polar

## Checklist
**Before PR**:

- [ ] I have read and followed the workflow indicated in the [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: adaptor/codegen/gen.py [hunk 1]
```diff
@@ -38,14 +41,18 @@
...
+    'complex32': 'complex64',
+    'complex128': 'complex64'
 }
 
 cast_strategy = {
     'NoCast': {},
     'Default': {
         'int64': 'int32',
         'float64': 'float32',
+        'complex32': 'complex64',
```
Comment (yangbofun) [inline adaptor/codegen/gen.py:54]: 这一条删掉

File: impl/camb/common/dtype_cast.cpp [hunk 2]
```diff
@@ -8,14 +8,36 @@
...
+//     {diopi_dtype_complex128, diopi_dtype_float64},
+//     {diopi_dtype_complex64, diopi_dtype_float32},
+//     {diopi_dtype_complex32, diopi_dtype_float16},
+// };
+
+diopiDtype_t getComplexMemberDtype(diopiDtype_t complexDtype) {
+    switch (complexDtype) {
+        case diopi_dtype_complex128:
+            return diopi_dtype_float64;
+            break;
```
Comment (yangbofun) [inline impl/camb/common/dtype_cast.cpp:31]: 删掉break，下同

File: impl/camb/common/dtype_cast.cpp [hunk 3]
```diff
@@ -8,14 +8,36 @@
...
+        case diopi_dtype_complex64:
+            return diopi_dtype_float32;
+            break;
+        case diopi_dtype_complex32:
+            return diopi_dtype_float16;
```
Comment (yangbofun) [inline impl/camb/common/dtype_cast.cpp:36]: 添加default，并设置set_last_error

File: impl/camb/common/dtype_cast.cpp [hunk 4]
```diff
@@ -96,6 +118,25 @@ diopiError_t dataTypeCast(diopiContextHandle_t ctx, DiopiTensor& dest, const Dio
...
+    CnnlDataType::convertToCnnlType(&srcCnnlDtype, srcDtype);
+    CnnlDataType::convertToCnnlType(&destCnnlDtype, destDtype);
+
+    // camb only support CNNL_DTYPE_COMPLEX_HALF and CNNL_DTYPE_FLOAT so far
+    if (CnnlDataType::isComplex(srcCnnlDtype) && CnnlDataType::isComplex(destCnnlDtype)) {
+        diopiDtype_t srcMemberType = getComplexMemberDtype(src.dtype());
+        diopiDtype_t destMemberType = getComplexMemberDtype(dest.dtype());
+        std::vector<int64_t> tempShape = src.shape();
+        tempShape.back() *= 2;
+        DiopiTensor tempSrcTensor = requiresTensor(ctx, tempShape, srcMemberType);
```
Comment (yangbofun) [inline impl/camb/common/dtype_cast.cpp:132]: 用原来的src直接转化就行，不用多copy一份src

File: impl/camb/convert_config.yaml [hunk 5]
```diff
@@ -1,5 +1,5 @@
 - common_config:
-    dtype: (int64)->int32, (float64)->float32
+    dtype: (int64)->int32, (float64)->float32, (complex32)->complex64, (complex128)->complex64
```
Comment (yangbofun) [inline impl/camb/convert_config.yaml:2]: 删除complex32->complex64

File: adaptor/codegen/gen.py (hunk 1 above)
Comment (NeosZhang) [reply adaptor/codegen/gen.py:54]: done

File: impl/camb/common/dtype_cast.cpp (hunk 2 above)
Comment (NeosZhang) [reply impl/camb/common/dtype_cast.cpp:31]: done

File: impl/camb/common/dtype_cast.cpp (hunk 3 above)
Comment (NeosZhang) [reply impl/camb/common/dtype_cast.cpp:36]: done

File: impl/camb/convert_config.yaml (hunk 5 above)
Comment (NeosZhang) [reply impl/camb/convert_config.yaml:2]: done

Comment (NeosZhang) [reply impl/camb/convert_config.yaml:2]: 寒武纪只支持complex32和complex64

**Review by yangbofun — APPROVED** [review]

**Pull request merged**